R/extend_scenario_trajectory.R

```diff
@@ -109,3 +108,177 @@ extend_scenario_trajectory <- function(data,
       .data$scenario_geography, .data$ald_sector, .data$technology, .data$year
     )
 }
+
+
+#' Extend the scenario pathways based on the fair share approach (now known as
+#' market share approach).  We first ensure that all company production plans
+#' are kept from the start year until the end of the forecast period.
+#' We then extend the scenario trajectories by multiplying the start value of
+#' the production with the relative change in each year/scenario combination.
+#' Contrary to the old version, this implements company targets based on the
+#' SMSP for increasing technologies and TMSR for decreasing ones.
+#' Companies that get production targets below 0 or that show a pattern of
+#' phasing out a technology within the forecast period, will get 0 scenario
+#' targets.
+#'
+#' @param data A data frame containing the production forecasts of companies
+#'   (in the portfolio). Pre-processed to fit analysis parameters and after
+#'   conversion of power capacity to generation.
+#' @param scenario_data A data frame containing scenario data for the specified
+#'   parameters of the analysis, including the business as usual and target
+#'   scenarios, the relevant scenario geography and time frame for each of the
+#'   technologies.
+#' @param start_analysis Numeric. A vector of length 1 indicating the start
+#'   year of the analysis.
+#' @param end_analysis Numeric. A vector of length 1 indicating the end
+#'   year of the analysis.
+#' @param time_frame Numeric. A vector of length 1 indicating the number of years
+#'   for which forward looking production data is considered.
+#' @noRd
+extend_scenario_trajectory <- function(data,
+                                       scenario_data = NULL,
+                                       start_analysis = NULL,
+                                       end_analysis = NULL,
+                                       time_frame = NULL) {
+  force(data)
+  scenario_data %||% stop("Must provide input for 'scenario_data'", call. = FALSE)
+  start_analysis %||% stop("Must provide input for 'start_analysis'", call. = FALSE)
+  end_analysis %||% stop("Must provide input for 'end_analysis'", call. = FALSE)
+  time_frame %||% stop("Must provide input for 'time_frame'", call. = FALSE)
+
+  validate_data_has_expected_cols(
+    data = data,
+    expected_columns = c(
+      "year", "investor_name", "portfolio_name", "equity_market", "ald_sector",
+      "technology", "scenario", "allocation", "scenario_geography",
+      "plan_tech_prod", "plan_carsten", "scen_tech_prod", "plan_sec_prod",
+      "plan_sec_carsten", "id", "company_name"
+    )
+  )
+
+  validate_data_has_expected_cols(
+    data = scenario_data,
+    expected_columns = c(
+      "source", "technology", "scenario_geography", "ald_sector", "units",
+      "scenario", "year", "direction", "fair_share_perc"
+    )
+  )
+
+  data <- data %>%
```

Comment:
it would really drastically imcrease the readibility of this function if you moved each block where at the end of a pipe u assign to data (and some calculation, not pure wrangling, happens) into its own function. As they are interna it would be ok to not document them (a good name is already helpful enough) so this should be doable very quickly.

Reply:
done

Reply:
will check the diffs tomorrow. The resulting files are rather similar in size though

Reply:
beautiful